### ISSUE TITLE
Generate interface for every service action

### ISSUE THREAD

**Issue by lukaszx0** [grpc/grpc-java#2446]
The generated service stubs evolved quite a lot in past few months and I've noticed that we removed service interface in favor of having just an abstract base class, it was indeed redundant and probably not very useful (I don't know if it was solve factor that led to removing it, if not please correct me).

While the per-service interface in the presence of equivalent abstract class might not provide a lot of value (in the context of generated services) I'd like to propose generating per-action interfaces which would be implemented by the base abstract service.

The value of per-action interfaces is that it would allow users to easily build custom service definitions which then would allow custom integrations for frameworks where single action is represented by class (as opposed to one class implementing all actions). This approach is used in many custom frameworks which relay heavily on dependency injection and the main benefit it provides is that in order to test single action, you don't need to provide dependencies for every action of a given service.

I'm open to work on implementation if we reach consensus. The change will be additive and won't break any existing APIs or implementations.

**Comment by daicoden** [comment]
This would be really helpful. Something like: https://github.com/daicoden/grpc-reflection-request/pull/1/files

It's not just tests but also resource scoping.

As an example, a service with different permissions (like read/write) might have access to different bound resources. It's hard to explain with a straight face to other developers why there are two optional transactors available (read & write) and how, based on scoping at dispatch time, one of the transactors is present, and the other is not, and that they should read the proto definition in order to figure out which one to use.

**Comment by ejona86** [comment]
It was removed because it is not API-safe. Interfaces can not receive new methods over time, whereas gRPC services can. If a new method is added to a gRPC service, it can break compilation of Java code which was deemed unacceptable (we can't just trust our users to "do the right thing" and avoid the interface when they don't own the service). https://github.com/grpc/grpc-java/issues/1469#issuecomment-226873511 had the discussion.

This is a case where a Java 8-specific API could do better, because it could provide default implementations. But it would probably need to be a separate API; it seems we'll need to provide Java 7 compatibility for a long time still (if for no other reason, Android). But such a thing causes trouble for toolchains, since we want to be able to provide pre-generated code to consumers. It's unclear how best to proceed and if it is worthwhile.

Frameworks where a single action is represented as a class should have no need to use the generated code (mostly; it may want the `MethodDescriptor`s/`ServiceDescriptor`). They should just create the `ServerServiceDefinition` "manually", reusing `ServerCalls` as helpful. I'm aware of such a framework implementation, so I'd be surprised if it was overly difficult. There is one oddity, that the implementation would generally create a new `ServerCallHandler` each RPC to have access to `ServerCall`, but that seems minor (i.e., `ServerCallHandler.asyncUnaryCall(invoker).startCall(call, headers);`).

**Comment by lukaszx0** [comment]
@ejona86 I'm not sure we're on the same page. I agree that per-service interfaces are problematic, but with our current `abstract ServiceImplBase` and default implementations for service methods, per-action interfaces won't introduce this problem.

The proposal is to add interfaces with one method representing single action and implement all of the interfaces it by `ServiceImplBase` which provides default "public" thing to use. Per-action interfaces wouldn't be used anywhere outside of generated file. They'd just provide easier way to implement per-action way of handling RPC in the aforementioned frameworks and custom ServerServiceDefinition setups. They won't break your stuff if you'll be relying on `ServiceImplBase` because it'll provide default `UNIMPLEMENTED` impls.

Also, if you're doing something custom like using building blocks for `ServiceDefinition` that we expose in generated files, you're on your own and should take the risk just like when using `METHOD_FOOBAR` constants etc. Again, the risk is very low because you'll be only affected if the given service method will be removed or changed... which should always happen in a coordinated way anyways (as well as in `ServiceImplBase` case).

> Frameworks where a single action is represented as a class should have no need to use the generated code (mostly; it may want the MethodDescriptors/ServiceDescriptor). They should just create the ServerServiceDefinition "manually", reusing ServerCalls as helpful. I'm aware of such a framework implementation, so I'd be surprised if it was overly difficult. There is one oddity, that the implementation would generally create a new ServerCallHandler each RPC to have access to ServerCall, but that seems minor (i.e., ServerCallHandler.asyncUnaryCall(invoker).startCall(call, headers);).

That's true and that's what we're currently doing. Adding per-action interfaces would just make it simpler.

It's additive, won't break existing stuff in any way, and doesn't carry a lot of weight in my opinion.

**Comment by lukaszx0** [comment]
TLDR: If somebody will use per-action interfaces, we don't have to provide default impl (which requires java 8) because you should simply just not "enable" those actions in your application or provide default impl by hand.

**Comment by carl-mastrangelo** [comment]
@lukaszx0 could you clarify what you mean by action?  Do you mean method?

**Comment by lukaszx0** [comment]
@carl-mastrangelo Yes. I mean service method (sorry, I use this naming interchangeably). @daicoden [diff shows what I'm talking about](https://github.com/daicoden/grpc-reflection-request/pull/1/files#diff-a7cfbc11f1e9a09de3c32e9e1366b64aR74)

**Comment by daicoden** [comment]
Thanks @lukaszx0 and @ejona86 for taking a look.  I did see #1479 before, and I definitely appreciate preventing breaking the build from additions.  In a large project it really did suck whenever someone would add a method, and the build would break with a compilation error.  

> They should just create the ServerServiceDefinition "manually", reusing ServerCalls as helpful. I'm aware of such a framework implementation, so I'd be surprised if it was overly difficult.

To add some color to the difficulty implementing the per method classes...

Right now to tell someone how to implement a method:
* implement one of 4 interfaces `ServerCalls.UnaryMethod<ReqT, RespT>`, `ServerCalls.ServerStreamingMethod<ReqT, RespT>` ...
* Register an ActionModule (custom service wiring)
* `bindAction(implementation, proto service name, proto method name)` - might be able to do something with the generated `METHODID_<name>`... but still doesn't tell you what service it's part of.

It feels very loose, and not like you are using GRPC properly.  When you read documentation about GRPC it seems like there must be something wrong with the framework because of how messy it is to get all the information needed for the custom service builder.

If every method creates an interface in the service namespace `interface Service.Method`, the implementation story would be:
* implement the method interface
* Register an `ActionModule` (custom service wiring).
* `bindAction(implementationClass)`

The wiring is then done by the framework without needing all the context of how to link the code together.  Reflection is used to figure out which methods go where.  There's still a little bit of configuration to say "this applications implements these rpc services" but it cleans it up a lot.

**Comment by ejona86** [comment]
@lukaszx0, okay, I see more now. I'm closer to the same page. I looked at @daicoden's diff, but missed what was pointed out (in part due to the implementation in the interfaces :) ). I see there aren't major breakage concerns (although we may want to choose interfaces names to prevent name collisions).

If I make a class that implements `CreateSession` (as a service author), how does it get used? The framework consumes it? And that's easier because now there is a concrete interface that applications can implement vs relying on annotations at runtime?

@daicoden wrote:
> As an example, a service with different permissions (like read/write) might have access to different bound resources. It's hard to explain to other developers why there are two optional transactors available (read & write) but they can only use one or the other depending on which method was invoked.

This is exposing a lot of your specific injection terminology/setup, so let me say what I think you're saying (and this is the first I've come across transactors). "Bound resources" is referring to dependencies being injected into a service method. Read and write may be two different resources injected. If all service methods are in a single class, then you can't prevent read-only service methods from receiving the write resource. And you do some verification (or wish to) that read-only service methods don't use the write resource.

> Would be nice to remove the optional injection and provide an application test that could alert about injection constraints.

"optional injection" seems more specific implementation details, but I think I follow the gist of wanting to verify in a test.

So all this injection stuff seems to just be why you'd "put each service method in a separate class." Is that true? If so, I'm fine with that part. I even see reason for instantiating a service method's class each RPC.

**Comment by daicoden** [comment]
@ejona86 That's pretty much exactly right.  Sorry for the specifics and thanks for the time. 

>  I see there aren't major breakage concerns (although we may want to choose interfaces names to prevent name collisions).

It might just work if interfaces were name spaced under the generated GRPC service.  Since two proto method's can't have the same name (protoc fails).

> If I make a class that implements CreateSession (as a service author), how does it get used? The framework consumes it? And that's easier because now there is a concrete interface that applications can implement vs relying on annotations at runtime?

The other change that would be nice for the framework would be for the method descriptor (or service descriptor) to have access to the interfaces that should be implemented by the service.

---

More context if needed,

I am actually working on making the RPC method classes instantiated on each RPC.  My main request is to figure out a way to make it easier for a developer to implement method class based RPCs.  Other suggestions welcome, but here's what I've got.

```
....
@GrpcService("edu.hawaii.cfht.protos.authorization")
class CreateSessionAction implements ServerCalls.UnaryMethod<CreateSessionRequest, CreateSessionResponse> {
  @Inject public CreateSessionAction(Provider<RequestScopedResource> resource) {...};

  @GrpcMethod("createSession") public void invoke(CreateSessionRequest request, io.grpc.stub.StreamObserver<CreateSessionResponse> responseObserver) {
    ...
  }
}
```

It works ok... but the framework code is very ugly, and, to the implementer, it feels unhinged from GRPC.  Any thoughts on how to clean this up with existing capabilities welcome!  But would love to get some better first class support from the generated GRPC code.

The best end state for me would be if the example code above had no mention of my framework.  The configuration of my framework could be handled completely separate.  Ideally, there would be enough reflective capabilities from method descriptors, service descriptors, and marker interfaces that I can figure out what a given rpc method is, and dispatch to it when appropriate.

**Comment by ejona86** [comment]
That GrpcService looks like the protobuf path. I think I'd have it be the generated class like `@GrpcService(AuthorizationGrpc.class)`. Then the framework code could easily call `AuthorizationGrpc.getServiceDescriptor()`.

In master, the protobuf marshaller is a `MethodDescriptor.PrototypeMarshaller` (you can cast) which has `Class<T> getMessageClass()` and `T getMessagePrototype()` methods. `ServiceDescriptor` also has `getMarshallerDescriptor()` which for protobuf (and casting) can provide you with a protobuf `FileDescriptor`. That should aid you greatly.

> for the method descriptor (or service descriptor) to have access to the interfaces that should be implemented by the service.

That's not a direction I feel comfortable with, because the descriptor shouldn't be concerned with the API used to implement the service. Yes, we only have async today, but blocking would be really nice. Or "real" Rx.

**Comment by lukaszx0** [comment]
> > The other change that would be nice for the framework would be for the method descriptor (or service descriptor) to have access to the interfaces that should be implemented by the service.
>
> That's not a direction I feel comfortable with, because the descriptor shouldn't be concerned with the API used to implement the service. Yes, we only have async today, but blocking would be really nice. Or "real" Rx.

Yeah, but returning _interfaces_ from `ServiceDescriptor` is really not needed and I agree with @ejona86 that we don't want to do this. We already have everything we need in descriptors to infer method interfaces from it - the list of the methods with its `MethodDescriptors`. With very simple reflection you'll be able to have small helpers returning interface names for the methods. But we probably won't even need that if we add support for it in `ServerServiceDefinition`.

Regarding the code that @daicoden pasted - I'd design it differently and we're doing it in a different way too. The whole point of interfaces is to get the method signature so you can implement it in your "action class" (aka service method class), so should do just this:

```@GrpcServiceAction(AuthenticationServiceGrpc.class)
class CreateSessionAction implements AuthenticationServiceGrpc.CreateSession<CreateSessionRequest, CreateSessionResponse> {
  @Inject public CreateSessionAction(Provider<RequestScopedResource> resource) {...};

      // method name should be the same as method name in service proto as well as the method interface that we're implementing
      public void createSession(CreateSessionRequest request, io.grpc.stub.StreamObserver<CreateSessionResponse> responseObserver) {
  }
}```.

In our framework, we do something a bit different, we define the RPC actions that are consumed/exposed by the app like this (the code is adjusted to what we could be doing with the "method interfaces" feature in place):

```@App("foo")
@ExposedServices({
    @Service(service = FooServiceGrpc.class, handlers = {
        // set class handler for the service method
        @Handler(method = FooServiceGrpc.Bar.class, handler = FooAction.class),
    }),
})
public class FooApp {
  public static void main(String[] args) throws Exception {
    AppContainer.forApp(FooApp.class).runAndExit(args);
  }
}
```

In general I think it would be nice if we could provide better, first class, support for per-method handlers as opposed to requiring implementing whole service class (`ServiceImplBase`) or requiring you to build this by hand. Adding method interfaces would be a first step, then we could add support for this in `ServerServiceDefinition` builder which would remove the need to be familiar the "internals" of generated stubs to build service definitions customly. @daicoden at first didn't even realized it's possible to do this until I explained to him how we're doing it internally.

I'll try to work on PR for interfaces today evening and we can continue iterating on this idea from there.

**Comment by daicoden** [comment]
Cool, great!  

@lukaszx0 the code I pasted is what I have now, would much prefer what you propose, so if you get a PR out I'll definitely use it!

**Comment by daicoden** [comment]
@lukaszx0 

Regarding:

```
@GrpcServiceAction(AuthenticationServiceGrpc.class)
class CreateSessionAction implements AuthenticationServiceGrpc.CreateSession<CreateSessionRequest, CreateSessionResponse> {
  @Inject public CreateSessionAction(Provider<RequestScopedResource> resource) {...};

      // method name should be the same as method name in service proto as well as the method interface that we're implementing
      public void createSession(CreateSessionRequest request, io.grpc.stub.StreamObserver<CreateSessionResponse> responseObserver) {
  }
}
```

Is there a reason the interface should be a template with ReqT and ResT?  Seems that it would be an error to choose the wrong req, resp, so we could simplify by just making the interface AuthenticationServiceGrpc.CreateSession and avoid errors (& error checking/messaging for the developer!)

**Comment by lukaszx0** [comment]
Yeah, might not be needed. I've just ad-hoc pseudo-coded that in the comment box without double checking.

**Comment by daicoden** [comment]
@lukaszx0 any movement on this?  If not I can put together a PR.

**Comment by ejona86** [comment]
I don't we were very excited about providing an interface for every service action. That gets pretty deep into the exact service API you are using. If I thought that multiple different DI frameworks would use it, it be easier to convince me, but in the past it seems every DI framework does things its own way.

Instead of just having this sit in limbo, I'm going to close it. The annotation piece of this can continue in #3173.
